core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1535,6 +1547,45 @@ class KafkaController(val config: KafkaConfig, zkClient: KafkaZkClient, time: Ti
 
 }
 
+case class ReassignmentStep(currentReplicas: Seq[Int], drop: Seq[Int], add: Option[Int], targetReplicas: Seq[Int]) {
+  override def toString: String = {
+    val currentFragment = currentReplicas.mkString("current replicas=", ","," ")
+    val dropFragment = if (drop.isEmpty) "" else drop.mkString("-", ","," ")
+    val addFragment = if (add.isEmpty) "" else s"+${add.get} "
+    val targetFragment = targetReplicas.mkString("target replicas=", ",", "")
+    s"$currentFragment$dropFragment$addFragment$targetFragment"
+  }
+}
+
+private[controller] object ReassignmentHelper {
+
+  // Rules
+  // - keep current leader as long as possible to avoid unnecessary leader elections.
+  // - in the last step order should match the requested order
+  def calculateReassignmentStep(finalTargetReplicas: Seq[Int], currentReplicas: Seq[Int], leader: Int) = {
+    val drop: Seq[Int] = calculateExcessISRs(finalTargetReplicas, currentReplicas, leader)
+    val add = finalTargetReplicas.filterNot(currentReplicas.contains).headOption
```

Comment:
An interesting thought here: if we won't take just the head but more then we can make this configurable. The other en, (take all) would actually be the non-incremental current behavior.

Reply:
Yeah, it probably should be configurable. In some cases, you need to optimize for the latency of the reassignment. For example, a user might have only a short window of low load during which they need to complete the rebalance.

Reply:
If you don't mind I'd separate this idea off this current PR as it probably needs a KIP and might delay this task.